docs/ref/contrib/admin/actions.txt

```diff
@@ -395,12 +402,48 @@ For example::
     class ArticleAdmin(admin.ModelAdmin):
         actions = ['make_published']
 
+        @admin.action(permissions=['publish'])
         def make_published(self, request, queryset):
             queryset.update(status='p')
-        make_published.allowed_permissions = ('publish',)
 
         def has_publish_permission(self, request):
             """Does the user have the publish permission?"""
             opts = self.opts
             codename = get_permission_codename('publish', opts)
             return request.user.has_perm('%s.%s' % (opts.app_label, codename))
+
+The ``action`` decorator
+========================
+
+.. function:: action(*, permissions=None, description=None)
+
+    .. versionadded:: 3.2
+
+    This decorator can be used for setting specific attributes on custom action
+    functions that can be used with
+    :attr:`~django.contrib.admin.ModelAdmin.actions`::
+
+        @admin.action(
+            permissions=['publish'],
+            description='Mark selected stories as published',
+        )
+        def make_published(self, request, queryset):
+            queryset.update(status='p')
+
+    This is equivalent to setting some attributes (with the original, longer
+    names) on the function directly::
+
+        def make_published(self, request, queryset):
+            queryset.update(status='p')
+        make_published.allowed_permissions = ['publish']
+        make_published.short_description = 'Mark selected stories as published'
```

Comment:
So if I come to this documentation (or for `@display`) for the first time, it tells me about setting these attributes but not what they mean, and nor is there an obvious link to where to read about them. It makes it a bit hard to use `@action` or `@display` - it still feels like I have to know about both ways.

Perhaps the decorator reference should become the canonical place where the meanings of the arguments and attributes are documented? At least, they can be briefly described with links to "more info" sections below, with each of the examples. The current structure is a fairly messy since it seems to have grown organically.

For actions:

* `short_description` is under "Writing action functions" only
* `allowed_permissions` has a whole section "Setting permissions for actions"

For display, everything is listed in bullet points under the `ModelAdmin.list_display` documentation (a giant list that starts "A few special cases"):

* `boolean` - this unnecessarily re-spells what kinds of strings and functions are allowed in `list_display`
* `admin_order_field`
* `short_description` - this only has one sentence saying "As some examples have already demonstrated" which doesn't feel very easy to use as a reference since it's saying "go read above"
* `empty_value_display` - this goes off on a bit of a tangent documenting the ability to change it for the whole admin site and the model admin class too. It would probably be best as its own section that documentation for the other two classes link to.

I am suggesting a fair amount of docs rearrangement here. I am happy to shunt it to a second PR and help on that.

Reply:
I agree. It is quite impenetrable and definitely needs rejigging. I tried not to move stuff around too much at this stage, however, and kept everything in the same place.

I think a follow up PR to improve the whole of this section would be a good idea - the large bulleted list doesn't feel the best way to present the options.